rust/src/pet/coordinator.rs

```diff
@@ -1,887 +1,314 @@
 #![allow(dead_code)] // temporary
 
-use std::{collections::HashMap, default::Default, iter::Iterator, ops::Range};
+use std::{collections::HashMap, default::Default};
 
+use counter::Counter;
 use sodiumoxide::{
     self,
-    crypto::{box_, sealedbox, sign},
+    crypto::{box_, hash::sha256, sign},
     randombytes::randombytes,
 };
 
-use super::{utils::is_eligible, PetError};
+use super::{
+    message::{Sum2Message, SumMessage, UpdateMessage},
+    utils::is_eligible,
+    PetError,
+};
+
+#[derive(Debug, PartialEq)]
+/// Round phases of a coordinator.
+enum Phase {
+    Idle,
+    Sum,
+    Update,
+    Sum2,
+}
 
 /// A coordinator in the PET protocol layer.
 pub struct Coordinator {
     // credentials
-    pub encr_pk: box_::PublicKey, // 32 bytes
-    encr_sk: box_::SecretKey,     // 32 bytes
+    encr_pk: box_::PublicKey, // 32 bytes
+    encr_sk: box_::SecretKey, // 32 bytes
+    sign_pk: sign::PublicKey, // 32 bytes
+    sign_sk: sign::SecretKey, // 64 bytes
 
     // round parameters
     sum: f64,
     update: f64,
     seed: Vec<u8>, // 32 bytes
+    min_sum: usize,
+    min_update: usize,
+    phase: Phase,
 
     // dictionaries
     dict_sum: HashMap<box_::PublicKey, box_::PublicKey>,
     dict_seed: HashMap<box_::PublicKey, HashMap<box_::PublicKey, Vec<u8>>>,
-}
-
-impl Coordinator {
-    pub fn new() -> Result<Self, PetError> {
-        // crucial: init must be called before anything else in this module
-        sodiumoxide::init()
-            .and(Ok(Default::default()))
-            .or(Err(PetError::InsufficientSystemEntropy))
-    }
-
-    /// Validate and handle a message.
-    pub fn validate_message(&self, message: &[u8]) -> Result<usize, PetError> {
-        if SumMessage::validate(self, message).is_ok() {
-            // placeholder: handle result
-            Ok(1)
-        } else if UpdateMessage::validate(self, message).is_ok() {
-            // placeholder: handle result
-            Ok(2)
-        } else if Sum2Message::validate(self, message).is_ok() {
-            // placeholder: handle result
-            Ok(3)
-        } else {
-            // unknown message type
-            Err(PetError::InvalidMessage)
-        }
-    }
+    masks: Counter<Vec<u8>>,
 }
 
 impl Default for Coordinator {
     fn default() -> Self {
```

Comment:
In Rust there's an implicit convention that `Default::default` always returns the same thing. This implementation breaks this convention because `seed`, and `{encr,sign}_{pk,sk}` are random. I think it would be best to move this initialization into a `new()` method instead.

Unrelated nit: I think a shorter version would be slightly more readable.

```rust
    fn default() -> Self {
        let (encr_pk, encr_sk) = box_::gen_keypair();
        let (sign_pk, sign_sk) = sign::gen_keypair();
        Self {
            encr_pk,
            encr_sk,
            sign_pk,
            sign_sk,
            sum: 0.01,
            update: 0.1,
            seed: randombytes(32),
            min_sum: 1,
            min_update: 3,
            phase: Phase::Idle,
            dict_sum: HashMap::new(),
            dict_seed: HashMap::new(),
            masks: Counter::new(),
        }
    }
```

Reply:
i'll split that.